File: frontend/src/lib/purchaseOrder.ts
// 采购单前端类型 + API。对齐后端 schemas/purchase_order.py。
// 红线字段(unit_price / line_total / total_amount)后端对无 purchase:read_cost 者脱敏为 null;
// 前端只需把 null 渲染成「—」(见 formatCost),不做任何客户端隐藏逻辑。
import { api } from "./api";
import type { Page } from "./catalog";
import { formatMoney } from "./format";
import { qs } from "./qs";

export type PurchaseOrderStatus = "DRAFT" | "CONFIRMED" | "CANCELLED";

/** 采购进度(销售单相对采购覆盖度的派生态)。 */
export type PurchaseProgress = "NOT_ORDERED" | "PARTIALLY_ORDERED" | "FULLY_ORDERED";

/** 收货进度(采购单相对入库覆盖度的派生态;入库步为 PO 详情/列表新增)。 */
export type ReceiptProgress = "NOT_RECEIVED" | "PARTIALLY_RECEIVED" | "FULLY_RECEIVED";

/** PO 详情「入库记录区」项(含 CANCELLED 并标状态,可追溯)。无成本字段。 */
export interface RelatedInboundOrder {
  id: number;
  no: string;
  status: "IN_TRANSIT" | "RECEIVED" | "CANCELLED";
  carrier_name: string | null;
  tracking_no: string | null;
  eta: string | null;
  arrived_at: string | null;
  created_at: string;
}

export interface PurchaseOrderOut {
  id: number;
  no: string;
  source_sales_order_id: number;
  supplier_id: number;
  currency: string;
  status: PurchaseOrderStatus;
  /** 红线:无 read_cost 时后端下发 null。 */
  total_amount: number | string | null;
  remark: string | null;
  updated_at: string;
  supplier_display?: string;
  source_sales_order_no?: string | null;
  /** 收货进度(入库步新增,详情响应附带)。 */
  receipt_progress?: ReceiptProgress;
}

export interface PurchaseOrderLineOut {
  id: number;
  purchase_order_id: number;
  sku_id: number;
  source_sales_order_line_id: number;
  name_snapshot: string;
  spec_text_snapshot: string;
  unit_snapshot: string;
  /** 红线:无 read_cost 时 null。 */
  unit_price: number | string | null;
  qty: number | string;
  /** 红线:无 read_cost 时 null。 */
  line_total: number | string | null;
  language: string;
  sort_order: number;
  remark: string | null;
  /** 入库覆盖度(入库步新增,详情响应附带每行)。 */
  received_qty?: number | string;
  in_transit_qty?: number | string;
}

export interface PurchaseOrderListItem {
  id: number;
  no: string;
  source_sales_order_id: number;
  source_sales_order_no: string | null;
  supplier_id: number;
  supplier_display: string;
  status: PurchaseOrderStatus;
  currency: string;
  /** 红线:无 read_cost 时 null。 */
  total_amount: number | string | null;
  line_count: number;
  created_at: string;
  /** 收货进度徽标(入库步新增,列表响应附带)。 */
  receipt_progress?: ReceiptProgress | null;
  /** 次要在途信号:已登记未收货的入库单数(与收货进度互补,回答「有几张货在路上」)。 */
  in_transit_count?: number;
}

/** 建单器可采行(GET /purchase-orders/purchasable-lines)。 */
export interface PurchasableLine {
  source_sales_order_line_id: number;
  sku_id: number;
  name_snapshot: string;
  spec_text_snapshot: string;
  unit_snapshot: string;
  required_qty: number | string;
  covered_qty: number | string;
  remaining_qty: number | string;
  /** 建议采购价(销售报价单价);无成本权限或缺省时 null。 */
  default_unit_price: number | string | null;
}

/** 写入行(id 有值=既有采购行对账更新;无 id=新增)。 */
export interface PurchaseOrderLineIn {
  id?: number;
  source_sales_order_line_id: number;
  qty: number;
  unit_price: number;
  remark?: string | null;
  sort_order?: number;
}

export interface PurchaseOrderCreateBody {
  source_sales_order_id: number;
  supplier_id: number;
  currency: string;
  remark?: string | null;
  lines: PurchaseOrderLineIn[];
}

/** PUT 必带乐观锁基线(= 上次 GET 的 order.updated_at)。 */
export interface PurchaseOrderUpdateBody {
  supplier_id: number;
  currency: string;
  remark?: string | null;
  lines: PurchaseOrderLineIn[];
  expected_updated_at: string;
}

/** 关联采购单(SO 详情附带,仅 purchase:read 者可见)。 */
export interface RelatedPurchaseOrder {
  id: number;
  no: string;
  status: PurchaseOrderStatus;
  supplier_id: number;
  supplier_display: string;
  currency: string;
  total_amount: number | string | null;
}

/** 红线金额渲染:null(无成本权限)→ 静音「—」,绝不显示 0。 */
export function formatCost(v: number | string | null | undefined): string {
  return v === null || v === undefined ? "—" : formatMoney(v);
}

export interface PurchaseOrderListFilters {
  status?: string;
  supplier_id?: number;
  source_sales_order_id?: number;
  source_sales_order_no?: string;
  /** 合并搜:采购单号 OR 来源销售单号(入库选单器用一个框)。 */
  q?: string;
  /** 仅列「还有未收量」的 PO,隐藏已全部入库的死单(入库选单器用)。 */
  receivable_only?: boolean;
  page?: number;
  size?: number;
}

export const purchaseOrderApi = {
  list: (p: PurchaseOrderListFilters) =>
    api.get<Page<PurchaseOrderListItem>>(`/api/v1/purchase-orders${qs(p as Record<string, unknown>)}`),
  get: (id: number) =>
    api.get<{ order: PurchaseOrderOut; lines: PurchaseOrderLineOut[]; inbound_orders: RelatedInboundOrder[] }>(
      `/api/v1/purchase-orders/${id}`,
    ),
  purchasableLines: (sourceSalesOrderId: number) =>
    api.get<{ items: PurchasableLine[] }>(
      `/api/v1/purchase-orders/purchasable-lines?source_sales_order_id=${sourceSalesOrderId}`,
    ),
  create: (b: PurchaseOrderCreateBody) =>
    api.post<{ order: PurchaseOrderOut; lines: PurchaseOrderLineOut[] }>("/api/v1/purchase-orders", b),
  update: (id: number, b: PurchaseOrderUpdateBody) =>
    api.put<{ order: PurchaseOrderOut; lines: PurchaseOrderLineOut[] }>(`/api/v1/purchase-orders/${id}`, b),
  del: (id: number) => api.del<null>(`/api/v1/purchase-orders/${id}`),
  confirm: (id: number) =>
    api.post<{ order: PurchaseOrderOut; lines: PurchaseOrderLineOut[] }>(`/api/v1/purchase-orders/${id}/confirm`),
  cancel: (id: number) =>
    api.post<{ order: PurchaseOrderOut; lines: PurchaseOrderLineOut[] }>(`/api/v1/purchase-orders/${id}/cancel`),
};

File: backend/app/schemas/purchase_order.py
"""采购单 schemas + 红线脱敏构造工厂。

🔴红线:采购价(unit_price)/ 行额(line_total)/ 单头金额(total_amount)对无 `purchase:read_cost`
者置 null。可见性下沉到 `.build(..., can_see_cost=...)` 构造工厂(单点经 redact_fields),
让「新出口忘记脱敏而裸露成本」在结构上更难发生,而非 dump 后 patch。
"""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.rbac.redaction import PO_COST_FIELDS, PO_LINE_COST_FIELDS, redact_fields
from app.schemas.common import LineQty, LineUnitPrice

# ---------- 写入(建单 / 编辑对账)----------


class PurchaseOrderLineIn(BaseModel):
    """建单行(无 id)。"""
    source_sales_order_line_id: int
    qty: LineQty
    unit_price: LineUnitPrice
    remark: str | None = None
    sort_order: int = Field(default=0, ge=0)


class PurchaseOrderCreateIn(BaseModel):
    """基于 SO 建一张 PO(单一供应商)。lines 允许空 → service 抛 41608(非 422)。"""
    source_sales_order_id: int
    supplier_id: int
    currency: str = Field(..., pattern=r"^[A-Z]{3}$")
    remark: str | None = None
    lines: list[PurchaseOrderLineIn] = Field(default_factory=list)


class PurchaseOrderLineEditIn(BaseModel):
    """编辑行:已存行带 id(对账 UPDATE),新行不带 id(INSERT)。"""
    id: int | None = None
    source_sales_order_line_id: int
    qty: LineQty
    unit_price: LineUnitPrice
    remark: str | None = None
    sort_order: int = Field(default=0, ge=0)


class PurchaseOrderUpdateIn(BaseModel):
    """整单保存(PUT):仅 DRAFT;必带乐观锁基线。source_sales_order_id 不可改(单据身份)。"""
    supplier_id: int
    currency: str = Field(..., pattern=r"^[A-Z]{3}$")
    remark: str | None = None
    lines: list[PurchaseOrderLineEditIn] = Field(default_factory=list)
    expected_updated_at: datetime


# ---------- 读投影(带红线脱敏工厂)----------


class PurchaseOrderLineOut(BaseModel):
    id: int
    purchase_order_id: int
    sku_id: int
    source_sales_order_line_id: int
    name_snapshot: str
    spec_text_snapshot: str
    unit_snapshot: str
    unit_price: float | None      # 🔴 无 read_cost → null
    qty: float
    line_total: float | None      # 🔴 无 read_cost → null
    language: str
    sort_order: int
    remark: str | None

    @classmethod
    def build(cls, line, *, can_see_cost: bool) -> dict:
        d = cls.model_validate(line, from_attributes=True).model_dump()
        return redact_fields(d, PO_LINE_COST_FIELDS, visible=can_see_cost)


class PurchaseOrderOut(BaseModel):
    id: int
    no: str
    source_sales_order_id: int
    supplier_id: int
    currency: str
    status: str
    total_amount: float | None    # 🔴 无 read_cost → null
    remark: str | None
    updated_at: datetime

    @classmethod
    def build(cls, po, extra: dict | None = None, *, can_see_cost: bool) -> dict:
        d = cls.model_validate(po, from_attributes=True).model_dump()
        d = redact_fields(d, PO_COST_FIELDS, visible=can_see_cost)
        if extra:
            d.update(extra)
        return d


class PurchaseOrderListItem(BaseModel):
    id: int
    no: str
    source_sales_order_id: int
    source_sales_order_no: str
    supplier_id: int
    supplier_display: str
    status: str
    currency: str
    total_amount: float | None    # 🔴 无 read_cost → null
    line_count: int
    created_at: datetime

    @classmethod
    def build(cls, item: dict, *, can_see_cost: bool) -> dict:
        d = cls.model_validate(item).model_dump()
        return redact_fields(d, PO_COST_FIELDS, visible=can_see_cost)


class RelatedPurchaseOrderItem(BaseModel):
    """SO 详情「关联采购单区」项(仅 purchase:read 下发)。含 CANCELLED 并标状态(可追溯)。
    金额红线:无 read_cost → null;供应商身份红线由端点级 purchase:read 门控(此结构本就不下发给 SALES)。"""
    id: int
    no: str
    status: str
    supplier_id: int
    supplier_display: str
    currency: str
    total_amount: float | None

    @classmethod
    def build(cls, item: dict, *, can_see_cost: bool) -> dict:
        d = cls.model_validate(item).model_dump()
        return redact_fields(d, PO_COST_FIELDS, visible=can_see_cost)
